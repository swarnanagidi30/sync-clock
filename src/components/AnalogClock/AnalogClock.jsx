import React, { useState, useEffect, useRef, useCallback } from 'react'
import './AnalogClock.scss';
import ClockEditor from '../ClockEditor';

export default function AnalogClock(props) {
    const { date, datediff, onDateChange } = props;
    const refClockInterval = useRef(null);

    const [hours, setHours] = useState(date.getHours());
    const [minutes, setMinutes] = useState(date.getMinutes());
    const [seconds, setSeconds] = useState(date.getSeconds());
    const [showEditor, setShowEditor] = useState(false)

    const handleDate = useCallback(
        () => {
            const formatTime = (time) => {
                return time < 10 ? `0${time}` : time;
            }
            date.setHours(date.getHours() + (datediff || 0));
            setHours(formatTime(date.getHours()));
            setMinutes(formatTime(date.getMinutes()));
            setSeconds(formatTime(date.getSeconds()));
        },
        [datediff, date],
    );

    useEffect(() => {

        refClockInterval.current = setInterval(handleDate, 1000);
        return () => {
            clearInterval(refClockInterval.current);
            console.log('cleared');
        }
    }, [datediff, refClockInterval, date, handleDate])

    const secondsDeg = seconds * 360 / 60;
    const minutesDeg = (minutes * 360 / 60) + (seconds * 360 / 60) / 60;
    const hoursDeg = hours * 360 / 12 + ((minutes * 360 / 60) / 12);

    const secondsStyle = {

        transform: `rotate(${secondsDeg}deg) translate(-50%, -50%)`
    };
    const minutesStyle = {
        transform: `rotate(${minutesDeg}deg) translate(-50%, -50%)`
    };
    const hoursStyle = {
        transform: `rotate(${hoursDeg}deg) translate(-50%, -50%)`
    };

    const handleEdit = () => {
        clearInterval(refClockInterval.current);
        setShowEditor(true);
    }
    const handleEditClose = (date) => {
        refClockInterval.current = setInterval(handleDate, 1000);
        setShowEditor(false);
        onDateChange(date);
    }
    return (
        <div className="analog-root">
            <div className={"analog-clock"}>
                <div className={"inner"}>
                    <div className={"hand second"} style={secondsStyle} />
                    <div className={"hand minute"} style={minutesStyle} />
                    <div className={"hand hour"} style={hoursStyle} />
                    <div className="graduations">
                        {[...Array(60)].map((x, i) =>
                            <div className="graduation" key={i} />
                        )}
                    </div>
                </div>
                <div className="clock-setting">
                    <i className="fa fa-cog" aria-hidden="true" onClick={handleEdit} ></i>
                </div>
                {showEditor && <ClockEditor date={date} onClose={handleEditClose} />}
            </div>
        </div>
    );
}
