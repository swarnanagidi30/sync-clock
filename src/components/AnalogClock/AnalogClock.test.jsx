import React from 'react';
import { render } from '@testing-library/react';
import AnalogClock from './AnalogClock';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
    const date = new Date(2020, 11, 20, 10, 50, 55);
    const tree = renderer
        .create(<AnalogClock date={date} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders Clocks Editor', () => {
    const date = new Date(2020, 11, 20, 10, 50, 55);
    const { container } = render(<AnalogClock date={date} />);
    expect(container.firstChild).toHaveAttribute('class');
});
