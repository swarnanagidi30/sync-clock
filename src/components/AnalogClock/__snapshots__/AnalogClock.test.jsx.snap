// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="analog-root"
>
  <div
    className="analog-clock"
  >
    <div
      className="inner"
    >
      <div
        className="hand second"
        style={
          Object {
            "transform": "rotate(330deg) translate(-50%, -50%)",
          }
        }
      />
      <div
        className="hand minute"
        style={
          Object {
            "transform": "rotate(305.5deg) translate(-50%, -50%)",
          }
        }
      />
      <div
        className="hand hour"
        style={
          Object {
            "transform": "rotate(325deg) translate(-50%, -50%)",
          }
        }
      />
      <div
        className="graduations"
      >
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
        <div
          className="graduation"
        />
      </div>
    </div>
    <div
      className="clock-setting"
    >
      <i
        aria-hidden="true"
        className="fa fa-cog"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
